registry/download.go

```diff
@@ -80,58 +91,53 @@ func DownloadPluginFromGithub(ctx context.Context, localPath string, org string,
 	return nil
 }
 
-func downloadFile(ctx context.Context, localPath string, url string) (err error) {
+func downloadFile(ctx context.Context, localPath string, urls ...string) (urlIndex int, err error) {
 	// Create the file
 	out, err := os.Create(localPath)
 	if err != nil {
-		return fmt.Errorf("failed to create file %s: %w", localPath, err)
+		return 0, fmt.Errorf("failed to create file %s: %w", localPath, err)
 	}
 	defer out.Close()
 
-	// Get the data
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
-	if err != nil {
-		return fmt.Errorf("failed create request %s: %w", url, err)
-	}
+	for r := 0; r < RetryAttempts; r++ {
+		for i, url := range urls {
+			// Get the data
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+			if err != nil {
+				return 0, fmt.Errorf("failed create request %s: %w", url, err)
+			}
 
-	err = retry.Do(
-		func() error {
 			// Do http request
 			resp, err := http.DefaultClient.Do(req)
 			if err != nil {
-				return fmt.Errorf("failed to get url %s: %w", url, err)
+				return 0, fmt.Errorf("failed to get url %s: %w", url, err)
 			}
-
 			// Check server response
-			if resp.StatusCode != http.StatusOK {
+			if resp.StatusCode == http.StatusNotFound && i < len(urls)-1 {
+				// check alternative url
+				resp.Body.Close()
+				continue
+			} else if resp.StatusCode != http.StatusOK {
 				fmt.Printf("Failed downloading %s with status code %d. Retrying\n", url, resp.StatusCode)
-				return fmt.Errorf("statusCode != 200")
+				resp.Body.Close()
+				break
 			}
-			defer resp.Body.Close()
 
 			fmt.Printf("Downloading %s\n", url)
 			bar := downloadProgressBar(resp.ContentLength, "Downloading")
 
 			// Writer the body to file
 			_, err = io.Copy(io.MultiWriter(out, bar), resp.Body)
 			if err != nil {
-				return fmt.Errorf("failed to copy body to file %s: %w", localPath, err)
+				return 0, fmt.Errorf("failed to copy body to file %s: %w", localPath, err)
 			}
-
-			return nil
-		},
-		retry.RetryIf(func(err error) bool {
-			return err.Error() == "statusCode != 200"
-		}),
-		retry.Attempts(RetryAttempts),
-		retry.Delay(RetryWaitTime),
-	)
-
-	if err != nil {
-		return fmt.Errorf("failed downloading: %s", url)
+			resp.Body.Close()
+			return i, nil
+		}
+		time.Sleep(RetryWaitTime)
```

Comment:
Why not keep retry logic and just don't retry on 404? Also with the logic above, looks like for each retry URLs would start from index 0?

Reply:
Until now we have retried on 404s, so I wanted to keep that behavior the same. But maybe we shouldn't 🤷 Though 404s can happen for other reasons, like the plugin version still being in the process of being published. So with this updated logic things are exactly the same as before, but if it's the cloudquery repo we also check the secondary url on 404 before sleeping for the next try.

Reply:
The retry logic was added due to https://github.com/cloudquery/plugin-sdk/issues/291 (503 GitHub errors).
I don't this we've see other error codes so I think we don't need to retry on 404

Reply:
Thanks for the context @erezrokah, I'll update this to not retry on 404s, but fall back if there are more urls to try.

Reply:
Updated 👍 I also added an integration test that tries out the various cases